src/commands/app/upgrade.tsx

```diff
@@ -227,3 +274,61 @@ async function forceTargetVersionSelection(
     targetAppVersionString,
   );
 }
+
+async function updateMissingSystemSoftwareDependency(
+  process: ProcessRenderer,
+  apiClient: MittwaldAPIV2Client,
+  appInstallationId: string,
+  dependency: AppSystemSoftwareDependency,
+  force: boolean,
+) {
+  const dependencySoftware = await apiClient.app.getSystemsoftware({
+    systemSoftwareId: dependency.systemSoftwareId,
+  });
+  assertSuccess(dependencySoftware);
+
+  const dependencyVersionList = await apiClient.app.listSystemsoftwareversions({
+    systemSoftwareId: dependency.systemSoftwareId,
+  });
+  assertSuccess(dependencyVersionList);
```

Comment:
This API endpoint natively supports a `versionRange` query parameter,[^1] which I'd recommend using. By relying on the backend for this, you're guaranteed to use the same version selection process that the mStudio uses.

Then there's also the `recommended` filter (about which you probably know more than me), which you could use to further filter the selection of system software versions (I don't know if it's possible for the endpoint to return ≠1 recommended versions, though).

[^1]: https://developer.mittwald.de/docs/v2/reference/app/app-list-systemsoftwareversions/

Reply:
Oh very good point, should've looked deeper into this.

The recommended filter will limit the returned software version to a certain amount of major, minor or patchlevel versions. For the example of PHP version it will return the latest three minor versions if they are within the version range.

Thus I'll leave the rest of this function as is, minus the semver check since that's redundant.